File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/rest/post/IncidentalPost.kt
package jp.co.toukei.log.trustar.rest.post

import jp.co.toukei.log.lib.jsonArr
import jp.co.toukei.log.lib.jsonArrNotNull
import jp.co.toukei.log.lib.util.RequestBodyJson
import jp.co.toukei.log.trustar.db.result.entity.CommonIncidentalHeaderResult
import jp.co.toukei.log.trustar.db.result.entity.CommonIncidentalTimeResult
import jp.co.toukei.log.trustar.db.result.entity.CommonUserSync
import jp.co.toukei.log.trustar.user.ClientInfo
import org.json.JSONObject
import third.jsonObj

class IncidentalPost(
        private val clientInfo: ClientInfo,
        @JvmField val headerList: List<CommonIncidentalHeaderResult>,
        @JvmField val timeList: List<CommonIncidentalTimeResult>
) : RequestBodyJson {

    private class R {
        @JvmField
        val header: MutableList<CommonIncidentalHeaderResult> = mutableListOf()
        @JvmField
        val time: MutableList<CommonIncidentalTimeResult> = mutableListOf()
    }

    private fun MutableMap<String, MutableMap<String, R>>.r(it: CommonUserSync): R {
        return getOrPut(it.companyCd) { mutableMapOf() }.getOrPut(it.userId, ::R)
    }

    override fun jsonBody(): JSONObject {
        val resultList: List<JSONObject> = run {
            mutableMapOf<String, MutableMap<String, R>>().apply {
                headerList.forEach { r(it).header += it }
                timeList.forEach { r(it).time += it }
            }.flatMap { (company, l) ->
                l.map { (user, r) ->
                    val map = r.time.groupByTo(LinkedHashMap()) { it.headerUUID }
                    val removedIncidental = mutableSetOf<String>()
                    val addedIncidental = mutableListOf<JSONObject>()
                    val editedIncidental = mutableListOf<JSONObject>()
                    val removedTime = mutableSetOf<JSONObject>()
                    val addedTime = mutableSetOf<JSONObject>()
                    val editedTime = mutableSetOf<JSONObject>()
                    val hg = r.header.groupBy { it.sheetNo == null }
                    hg[true]?.forEach { h ->
                        val times = map.remove(h.uuid)
                        if (!h.deleted) {
                            addedIncidental += jsonObj {
                                "guid" v h.uuid
                                "allocationNo" v h.allocationNo
                                "allocationRowNo" v h.allocationRowNo
                                "shipperCd" v h.shipperCd
                                "workCd" v h.workList.jsonArr()
                                "picId" v h.picId
                                "time" v times?.jsonArrNotNull { t ->
                                    if (t.deleted) null
                                    else jsonObj {
                                        "guid" v t.uuid
                                        "cls" v t.type
                                        "beginTime" v t.beginDatetime
                                        "endTime" v t.endDatetime
                                    }
                                }
                            }
                        }
                    }
                    hg[false]?.forEach { h ->
                        h.sheetNo?.let { s ->
                            if (h.deleted) {
                                map.remove(h.uuid)
                                removedIncidental += s
                            } else {
                                editedIncidental += jsonObj {
                                    "sheetNo" v s
                                    "shipperCd" v h.shipperCd
                                    "workCd" v h.workList.jsonArr()
                                    "picId" v h.picId
                                }
                            }
                        }
                    }
                    map.values.asSequence()
                            .flatten()
                            .forEach { t ->
                                val s = t.sheetNo
                                val n = t.sheetRowNo
                                if (s != null && n != null) {
                                    if (t.deleted) {
                                        removedTime += jsonObj {
                                            "sheetNo" v s
                                            "sheetRowNo" v n
                                        }
                                    } else {
                                        editedTime += jsonObj {
                                            "sheetNo" v s
                                            "sheetRowNo" v n
                                            "beginTime" v t.beginDatetime
                                            "endTime" v t.endDatetime
                                        }
                                    }
                                } else {
                                    addedTime += jsonObj {
                                        "guid" v t.uuid
                                        "sheetNo" v s
                                        "cls" v t.type
                                        "beginTime" v t.beginDatetime
                                        "endTime" v t.endDatetime
                                    }
                                }
                            }
                    jsonObj {
                        "companyCd" v company
                        "userId" v user
                        "removeIncidental" v removedIncidental.jsonArr()
                        "editIncidental" v editedIncidental.jsonArr()
                        "addIncidental" v addedIncidental.jsonArr()
                        "editTime" v editedTime.jsonArr()
                        "removeTime" v removedTime.jsonArr()
                        "addTime" v addedTime.jsonArr()
                    }
                }
            }
        }

        return jsonObj {
            "incidentalResults" v resultList.jsonArr()
            "clientInfo" v clientInfo.jsonBody()
        }
    }
}

File: LogiOS/TrustarApp/Services/Request/IncidentalPost.swift
//
//  IncidentalPost.swift
//  TrustarApp
//
//  Created by CuongNguyen on 08/03/2022.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation

class IncidentalPost {
    private var clientInfo: ClientInfo
    private var headerList: [CommonIncidentalHeaderResult]
    private var timeList: [CommonIncidentalTimeResult]
    
    init(clientInfo: ClientInfo,
         headerList: [CommonIncidentalHeaderResult],
         timeList: [CommonIncidentalTimeResult]
    ) {
        self.clientInfo = clientInfo
        self.headerList = headerList
        self.timeList = timeList
    }
    
    private class R {
        var header: [CommonIncidentalHeaderResult] = []
        var time: [CommonIncidentalTimeResult]     = []
    }
        
    private var r: [String : [String : R]] = [String : [String : R]]()
    
    class IncidentalResult: Codable {
        var companyCd: String                 = String()
        var userId: String                    = String()
        var removeIncidental: [String]        = []
        var editIncidental: [EditIncidental]  = []
        var addIncidental: [AddIncidental]    = []
        var editTime: [EditTime]              = []
        var removeTime: [RemoveTime]          = []
        var addTime: [AddTime]                = []
    }
    
    class EditIncidental: Codable {
        var sheetNo: String = String()
        var shipperCd: String = String()
        var workCd: [String] = []
        var picId: String? = nil
    }
    
    class AddIncidental: Codable {
        var guid: String = String()
        var allocationNo: String = String()
        var allocationRowNo: Int = 0
        var shipperCd: String = String()
        var workCd: [String] = []
        var picId: String? = nil
        var time: [Time] = []
        
        class Time: Codable {
            var guid: String = String()
            var cls: Int = 0
            var beginTime: Int64?
            var endTime: Int64?
        }
    }
    
    class EditTime: Codable {
        var sheetNo: String = String()
        var sheetRowNo: Int = 0
        var beginTime: Int64?
        var endTime: Int64?
    }
    
    class RemoveTime: Codable {
        var sheetNo: String = String()
        var sheetRowNo: Int = 0
    }
    
    class AddTime: Codable {
        var guid: String = String()
        var sheetNo: String = String()
        var cls: Int = 0
        var beginTime: Int64?
        var endTime: Int64?
    }
    
    class ResultModel: Codable {
        var incidentalResults: [IncidentalResult]
        var clientInfo: ClientInfo
        
        init(incidentalResult: [IncidentalResult], clientInfo: ClientInfo) {
            self.incidentalResults = incidentalResult
            self.clientInfo = clientInfo
        }
    }
    
    func result() -> ResultModel {
        r = [String : [String : R]]()
        
        // create dictionary:
        // If contain -> append
        // if not contain -> add new
        for it in headerList {
            if var valByCompany = r[it.companyCd] {
                if let valByUserId = valByCompany[it.userId] {
                    valByUserId.header.append(it)
                } else {
                    let obj = R()
                    obj.header.append(it)
                    valByCompany[it.userId] = obj
                }
            } else {
                var dicObj = [String : R]()
                let obj = R()
                obj.header.append(it)
                dicObj[it.userId] = obj
                r[it.companyCd] = dicObj
            }
        }
        
        for it in timeList {
            if var valByCompany = r[it.companyCd] {
                if let valByUserId = valByCompany[it.userId] {
                    valByUserId.time.append(it)
                } else {
                    let obj = R()
                    obj.time.append(it)
                    valByCompany[it.userId] = obj
                }
            } else {
                var dicObj = [String : R]()
                let obj = R()
                obj.time.append(it)
                dicObj[it.userId] = obj
                r[it.companyCd] = dicObj
            }
        }
        
        var result: [IncidentalResult] = []
        
        for l in r {
            var removedIncidental : [String]           = []
            var addedIncidental   : [AddIncidental]    = []
            var editedIncidental  : [EditIncidental]   = []
            var removedTime       : [RemoveTime]       = []
            var addedTime         : [AddTime]          = []
            var editedTime        : [EditTime]         = []
            
            let item: IncidentalResult = IncidentalResult()
            
            item.companyCd = l.key
            
            for r in l.value {
                item.userId = r.key
                var map = Dictionary(grouping: r.value.time, by: { $0.headerUUID })
                let hg = Dictionary(grouping: r.value.header, by: { $0.sheetNo == nil })
                
                if hg[true] != nil {
                    for h in hg[true]! {
                        let times = map.removeValue(forKey: h.uuid)
                        if !h.deleted {
                            let ai: AddIncidental = AddIncidental()
                            addedIncidental.append(ai)
                            ai.guid = h.uuid
                            ai.allocationNo = h.allocationNo
                            ai.allocationRowNo = h.allocationRowNo
                            ai.shipperCd = h.shipperCd
                            ai.workCd = h.workList.components(separatedBy: ",")
                            ai.picId = h.picId
                            
                            if times != nil {
                                for t in times! {
                                    if !t.deleted {
                                        let at: AddIncidental.Time = AddIncidental.Time()
                                        ai.time.append(at)
                                        at.guid = t.uuid
                                        at.cls = t.type
                                        at.beginTime = t.beginDatetime
                                        at.endTime = t.endDatetime
                                    }
                                }
                            }
                        }
                    }
                }
                
                if hg[false] != nil {
                    for h in hg[false]! {
                        if let s = h.sheetNo {
                            if h.deleted {
                                map.removeValue(forKey: h.uuid)
                                removedIncidental.append(s)
                            } else {
                                let ei: EditIncidental = EditIncidental()
                                editedIncidental.append(ei)
                                ei.sheetNo = s
                                ei.shipperCd = h.shipperCd
                                ei.workCd = h.workList.components(separatedBy: ",")
                                ei.picId = h.picId
                            }
                        }
                    }
                }
                
                for value in map.values {
                    for t in value {
                        let s = t.sheetNo
                        let n = t.sheetRowNo
                        if s != nil && n != nil {
                            if t.deleted {
                                let rt: RemoveTime = RemoveTime()
                                removedTime.append(rt)
                                rt.sheetNo = s!
                                rt.sheetRowNo = n!
                            } else {
                                 let et: EditTime = EditTime()
                                 editedTime.append(et)
                                 et.sheetNo = s!
                                 et.sheetRowNo = n!
                                 et.beginTime = t.beginDatetime
                                 et.endTime = t.endDatetime
                            }
                        } else {
                            let at: AddTime = AddTime()
                            addedTime.append(at)
                            at.guid = t.uuid
                            at.sheetNo = s!
                            at.cls = t.type
                            at.beginTime = t.beginDatetime
                            at.endTime = t.endDatetime
                        }
                    }
                }
            }
            
            item.removeIncidental = removedIncidental
            item.addIncidental = addedIncidental
            item.editIncidental = editedIncidental
            item.removeTime = removedTime
            item.addTime = addedTime
            item.editTime = editedTime
            
            result.append(item)
        }
        
        return ResultModel(incidentalResult: result, clientInfo: clientInfo)
    }
}
